### PR TITLE
#388 Forgot password.

### PR DESCRIPTION
**Pull request by ammaratef45**
PR for #388
  - Added `MAIL_DEFAULT_SENDER` to config
  - Added Flask-Mail
  - Made forgot password send a random password to the user's mail
  - Added new puzzle for testing the routine

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#440](https://github.com/timelesslounge/timelessis/issues/440) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#440](https://github.com/timelesslounge/timelessis/issues/440) is assigned to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), here is [why](https://www.0crat.com/footprint/CF9LG1YCA/0dd91416-2602-4b94-a9e6-f229ac8daa2c); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: timeless/auth/auth.py [hunk 1]
```diff
@@ -1,6 +1,14 @@
+""" Authentication methods are implemented here. """
+import random
+import string
+
 from flask import session
+from passlib.hash import bcrypt_sha256
 from timeless.employees.models import Employee
+from timeless.mail import MAIL
+from flask_mail import Message
```
Comment (vladarefiev) [inline timeless/auth/auth.py]: @ammaratef45 one more blank line pls

File: timeless/auth/auth.py [hunk 2]
```diff
@@ -24,10 +32,22 @@ def login(username="", password=""):
 
 
 def forgot_password(email=""):
+    """ Handle the forgot password routine. """
     user = Employee.query.filter_by(email=email).first()
```
Comment (vladarefiev) [inline timeless/auth/auth.py]: @ammaratef45 I'd like to see here fail first. 
```
if not email:
   return ...
```

File: timeless/auth/auth.py [hunk 3]
```diff
@@ -24,10 +32,22 @@ def login(username="", password=""):
...
     error = None
-    if user is None:
+    if not user:
```
Comment (vladarefiev) [inline timeless/auth/auth.py]: @ammaratef45 let's exit here
```
if not user:
    return ...
```
and no need to check `if not error`

File: timeless/mail/__init__.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+""" MAIL module """
+from flask_mail import Mail
+
+MAIL = Mail()
```
Comment (vladarefiev) [inline timeless/mail/__init__.py]: @ammaratef45 I'd like to disable it for development environment and tests. Let's make a puzzle for it.

File: timeless/mail/__init__.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+""" MAIL module """
+from flask_mail import Mail
+
```
Comment (vladarefiev) [inline timeless/mail/__init__.py]: @ammaratef45 one more blank line pls

File: timeless/auth/auth.py [hunk 6]
```diff
@@ -1,6 +1,14 @@
+""" Authentication methods are implemented here. """
+import random
+import string
+
```
Comment (vladarefiev) [inline timeless/auth/auth.py]: @ammaratef45 blank line pls

**Comment by vladarefiev** [comment]
@ammaratef45 Hi, I've just leave a couple comments.

File: timeless/auth/auth.py (hunk 2 above)
Comment (ammaratef45) [reply timeless/auth/auth.py]: @vladarefiev `not email` is never true because it takes a default value of "" if not provided.

**Comment by ammaratef45** [comment]
@vladarefiev  sorry for late response, please take a look.

File: timeless/auth/auth.py (hunk 2 above)
Comment (vladarefiev) [reply timeless/auth/auth.py]: @ammaratef45 
```
>>> email=""
>>> if not email:
...   print("email is emprt")
... 
email is emprt
```

Comment (vladarefiev) [reply timeless/auth/auth.py]: @ammaratef45 FYI

**Comment by vladarefiev** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by emilianodellacasa** [comment]
@ammaratef45 we have to fix conflicts

**Comment by ammaratef45** [comment]
@emilianodellacasa Done 👍

**Comment by emilianodellacasa** [comment]
@rultor merge

**Comment by 0crat** [comment]
The job [#440](https://github.com/timelesslounge/timelessis/issues/440) is now out of scope

**Comment by 0crat** [comment]
Job [#440](https://github.com/timelesslounge/timelessis/issues/440) is not in the agenda of @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), can't inspect

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @vladarefiev[/z](https://www.0crat.com/u/vladarefiev), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @emilianodellacasa[/z](https://www.0crat.com/u/emilianodellacasa)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @vladarefiev[/z](https://www.0crat.com/u/vladarefiev)

**Comment by 0crat** [comment]
Quality review completed: +6 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**